Add support for poetry v2

DBS-ST-VIT opened the pull request:
Hey @ajkerrigan,
this maybe solves #26.

Actually, we implemented it with legacy support for poetry 1.x.x, but thats something we can also remove. We thought, that it is may be more user friendly with some time with legacy support. Maybe we should also add a deprecation notice for usage with poetry 1.x.x?

Further, there was one deprecation ([see here](https://github.com/python-poetry/poetry-core/blob/main/src/poetry/core/packages/package.py#L274)) which has been fixed by actually adapting the current code from poetry. For us, this doesn't look like the best solution, but poetry-plugin-export is expecting a BaseMarker object, which is the reason why we need to work with the marker here and cannot use e.g. the `.python_constraint` property instead.

DBS-ST-VIT committed:
feat: add support for poetry v2

DBS-ST-VIT committed:
fix: add workaround for deprecation of python_marker in poetry.core.packages.package.Package

ajkerrigan commented on src/poetry_plugin_freeze/app.py:
```diff
@@ -22,10 +23,18 @@
 from poetry_plugin_export.walker import get_project_dependency_packages, walk_dependencies
 
 
+is_legacy_poetry = parse_semver(poetry.core.__version__) < parse_semver("2")
+
+if is_legacy_poetry:
+    from poetry.core.pyproject.exceptions import PyProjectException
```
This came from an [exception rename](https://github.com/python-poetry/poetry/pull/9729/files#diff-c7d904b3993aceb26aa850119fdae2cfb308fdfbddefba3e8fbd0d974371c872R94) on Poetry's end. We may be better off trying to import both forms, rather than trying to selectively import based on version:

```suggestion
try:
    from poetry.core.pyproject.exceptions import PyProjectError
except ImportError:
    from poetry.core.pyproject.exceptions import PyProjectException as PyProjectError
```

...which would let us catch `PyProjectError` directly below.

DBS-ST-VIT replied:
This is actually a good point, and i also applied it.

But there is one point, you need to know: in the situation, that we're going to capture, poetry 2.x is raising a RuntimeError instead of the new PyProjectError. I dont know why, but i stumbled above this pitfall, since i first also assumed, that it is like you described.

ajkerrigan replied:
Good point, I think we need to handle those 2 cases separately. One is the `PyProjectError` (formerly `PyProjectException`) which can come up in some cases (like https://github.com/cloud-custodian/poetry-plugin-freeze/issues/14).

Other issues can raise a `RuntimeError` like you mentioned. In either case we should skip freezing that wheel and move on.

ajkerrigan commented on src/poetry_plugin_freeze/app.py:
```diff
@@ -12,8 +12,9 @@
 
 from cleo.helpers import option
 from poetry.console.commands.command import Command
+import poetry.core
+from packaging.version import parse as parse_semver
```
I think with a different import approach (in a comment below) we can avoid the reliance on version sniffing.

```suggestion
```

ajkerrigan reviewed (changes requested):
I'm generally good with this, the v2 changes don't look bad at all. Your handling of the python_marker change looks good :+1: .

A couple other comments/suggestions inline related to the upstream rename of `PyProjectException` --> `PyProjectError` and how to gracefully handle that.

DBS-ST-VIT commented:
Hey @ajkerrigan,
many thanks for your input! I updated the branch in our forked repo according your suggestions and also cleaned a little bit up.

May you can have a look on it?

WhoAmI0501 commented:
Hi @ajkerrigan ,
i am also waiting for this PR for upgrading to poetry v2. This is the last poetry plugin in my collection that is lacking v2 support. Can you merge this PR?

RudolfVonKrugstein commented:
Oh, this would be cool! I want to upgrade to poetry V2, but need this plugin.

ajkerrigan committed:
skip freezing wheels in the case of PyProjectError or RuntimeError

Also handle the upstream PyProjectException-->PyProjectError rename.

ajkerrigan reviewed (approved):
I'm good with these changes, pushed one minor tweak to make sure we're covering the `PyProjectException`/`PyProjectError`/`RuntimeError` cases. Thanks for working on this one @DBS-ST-VIT :+1:
